src/main/java/wellnus/gamification/util/GamificationTokenizer.java

```diff
@@ -0,0 +1,58 @@
+package wellnus.gamification.util;
+
+import java.util.ArrayList;
+
+import wellnus.exception.TokenizerException;
+import wellnus.storage.Tokenizer;
+
+/**
+ * Handles the conversion of GamificationData objects -> String and vice versa to allow
+ * storage and retrieval of gamification statistics.
+ */
+public class GamificationTokenizer implements Tokenizer<GamificationData> {
+    private static final String INVALID_STORED_DATA_MESSAGE = "Invalid gamification data %s found in storage! "
+            + "Gamification data will not be restored";
+    /**
+     * Converts the attributes of the <code>GamificationManager</code> into a String representation to be
+     * saved to storage.
+     *
+     * @param dataObjects List of GamificationData Objects we want to convert into a String representation
+     * @return ArrayList of Strings representing the GamificationData objects that we can write to storage
+     */
+    @Override
+    public ArrayList<String> tokenize(ArrayList<GamificationData> dataObjects) {
+        ArrayList<String> tokenizedObjects = new ArrayList<>();
+        for (GamificationData data : dataObjects) {
+            int xp = data.getTotalXp();
+            String tokenizedObject = "" + xp;
+            tokenizedObjects.add(tokenizedObject);
+        }
+        return tokenizedObjects;
+    }
+
+    /**
+     * Converts the String representation of the <code>GamificationManager</code>'s state back into an
+     * <code>ArrayList</code> of GamificationData that can be used to restore the gamification feature's
+     * previous state.
+     *
+     * @param tokenizedDataObjects String representation of the GamificationData Objects whose state we want to restore
+     * @return ArrayList containing all the gamification data from the gamification feature's previously saved state
+     * @throws TokenizerException If detokenizing fails and stored gamification statistics cannot be restored
+     */
+    @Override
+    public ArrayList<GamificationData> detokenize(ArrayList<String> tokenizedDataObjects)
+            throws TokenizerException {
+        ArrayList<GamificationData> dataObjects = new ArrayList<>();
+        for (String tokenizedDataObject : tokenizedDataObjects) {
+            int totalXp;
+            try {
+                totalXp = Integer.parseInt(tokenizedDataObject.trim());
+            } catch (NumberFormatException numberFormatException) {
+                throw new TokenizerException(String.format(INVALID_STORED_DATA_MESSAGE, tokenizedDataObject));
+            }
+            GamificationData data = new GamificationData(totalXp);
+            dataObjects.add(data);
+        }
+        return dataObjects;
+    }
+}
```

Comment:
Before I approve this PR - can I verify what the behaviour is like if gamif is accessed (therefore the file is created) but nothing is saved?

Will there be a visible error on next `WellNUS++` load?

Reply:
Aha, this particular edge case is actually caught in `GamificationStorage` line 33. No exception will be thrown if data file doesn't exist. My intention is to _not_ create the gamif data file if no data is saved(i.e. no XP points updated during the app run), since any XP update will trigger a `Storage.saveData()` call that creates the gamif data file if it doesn't already exist.

Basically lazy initialisation of the data file to reduce unnecessary I/O.

Tested this to ensure that no exceptions occur in 4 cases:
1. gamif data file doesn't exist, XP updated: Stats is restored successfully on the next run of the app
2. gamif data file exists, XP updated: stats is updated and restored successfully the next run of the app
3. gamif data file doesn't exist, XP not updated: stats is the default value, no exceptions thrown, stats is still the default value on the next run of the app
4. gamif data file exists, XP not updated: stats is restored successfully

Reply:
Excellent!